02-intermediate-ruby/debugging.md

```diff
@@ -0,0 +1,108 @@
+# Debugging
+
+## Learning Goals
+- Read and analyze a stack trace
+- Understand how to use `pry` for debugging in Ruby code
+- Better understand the scope of variables and instances in an application when testing
+
+### Reading Errors via Stack Traces
+
+When an exception is raised, it immediately stops the current method, just like `return`. It will then bubble its way up through the program, method by method. If it makes it all the way to the top, Ruby will print out a summary of the exception and a description of what the program was doing when the exception happened, also known as a **stack trace**.
+
+A stack trace contains a ton of useful information, including a list of methods and blocks that Ruby was inside when the exception was raised. Being able to quickly read a complex stack trace is a super useful skill, so let's examine one now.
+
+First, take a look at the following stack trace that came from running some of the tests within the grocery store project.
+
+![test stack trace](images/test-stack-trace.png)
```

Comment:
Ruby 2.5 changed the format in which stack traces print. The order is reversed and they have a little more information. Does this image need to be updated?

Reply:
Hmm I guess I haven't upgraded to 2.5 so i'll do that and then see.

Reply:
Nope - I just ran it using rake on my project and this is how it looks. Is what you're thinking of just the generic stack trace, not an error from minitest?

Reply:
Probably. That's kind of annoying that they don't use the new (I would say more readable) format.